lib/ecto/schema.ex

```diff
@@ -2035,18 +2035,22 @@ defmodule Ecto.Schema do
     end
   end
 
-  defp check_field_type!(name, :datetime, _opts) do
+  defp check_field_type!(name, :datetime, _, _opts) do
     raise ArgumentError, "invalid type :datetime for field #{inspect name}. " <>
                            "You probably meant to choose one between :naive_datetime " <>
                            "(no time zone information) or :utc_datetime (time zone is set to UTC)"
   end
 
-  defp check_field_type!(name, type, opts) do
+  defp check_field_type!(name, type, full_type, opts) do
     cond do
       type == :any and !opts[:virtual] ->
         raise ArgumentError, "only virtual fields can have type :any, " <>
                              "invalid type for field #{inspect name}"
 
+      is_tuple(type) and tuple_size(type) == 2 and type |> elem(0) |> Ecto.Type.composite?() ->
```

Comment:
Ah, good point. Can we at least move this to a private function? so we just call it as `composite?(type) ->` or, even better, `inner_type = inner_from_composite(type) ->`. WDYT?